### ISSUE TITLE
Lua based property list

### ISSUE THREAD

**Issue by lvjr** [latex3/latex3#1040]
Recently Lua based `intarray`s were added to `l3kernel`. Are there any plans to add Lua based property lists? Currently, it is rather slow to read from and write to a `prop` with lots of items.

PS: Maybe in the far far future, eTeX engines would add property list as a new data structure.

**Comment by josephwright** [comment]
For large data sets, one can't get away from using a hash table based approach. The team have discussed some 'dynamic' data types that would use different implementations depending on the size, but this hasn't been a priority.

Specifically on using Lua, the issue is that we loose category codes. That's fine for the key but not the value. LuaTeX lets us restore catcodes using a catcode table, but we'd need to be sure to check every value to ensure it didn't need special treatment. I can imagine how that might work, but it's non-trivial. @zauguin might have some clever ideas, but I think it still needs an amount of macro code, at least at the point-of-storage.

**Comment by josephwright** [comment]
I very much doubt we'd want to look at new datatypes at the engine level - the hash table approach works well for e.g. `pgfplots` and really TeX isn't well-suited to handling large data sets at all.

**Comment by zauguin** [comment]
> Specifically on using Lua, the issue is that we loose category codes. That's fine for the key but not the value.

That's not a issue since we can scan the value as a list of tokens instead of as a string, then the category codes get preserved. That wouldn't really work for the keys since token lists don't play nice with hashing, but there we don't care about category codes anyway. Using token lists there is not exactly the fastest approach, but it shouldn't be that bad.

The bigger issue I see is about local property lists. Moving intarrays to Lua was relativly painless since they are always global anyway. Making Lua data structures respect TeX's grouping is rather hard. (aka. all reliable ways I know for that require LuaMetaTeX)

**Comment by blefloch** [comment]
To keep most operations O(1), you can use a doubly linked list: let's say the prop is called `\l_my_prop` and it has keys `A`, `B`, `C` with values `Alice`, `Bob`, `Carol`.  Then the structure could be (where `@` should be replaced by a string unique to each prop and kept short to lighten the load on the hash table). 
```
\tl_set:Nn \l_my_prop { \@_C { } \@_A }
\tl_set:Nn \@_A { \l_my_prop {Alice} \@_B }
\tl_set:Nn \@_B { \@_A {Boris} \@_C }
\tl_set:Nn \@_C { \@_B {Carol} \l_my_prop }
```
The forward pointers allow to map through the list.  The backwards pointers allow to delete an entry: if we want to delete the entry `\@_B` then we first expand it to see the previous and next entries, we replace `\@_B` to `\@_C` or `A` in the definitions of `A` and `C` respectively, and then we undefine `\@_B`.  To make the code as uniform as possible when removing the first or last entry, I made the structure above into a *circular* doubly-linked list, where `\l_my_prop` is an (empty) place in the doubly-linked list.  This is probably not optimal.

Presumably we could even make the whole shtick depend on how many items the prop has (converting from our usual prop to this system once it becomes large enough, but never converting back unless it becomes empty perhaps).  Alternatively we could introduce `\prop_large_new:N` defining a prop that uses this system, and then have the rest of the code easily distinguish between the two types of prop via some marker.

The main drawback is that people may have been `o`-expanding props and moving their data around, e.g. doing `\tl_set_eq:NN` on props and assuming that this is safe, whereas now it would start behaving strangely since the two resulting props would use the same underlying control sequences.  Another drawback is that it slows down mapping by a few additional expansions.  Overall I'm quite tempted, am I missing something?

**Comment by josephwright** [comment]
@blefloch  I like the idea - it's meant to be one of the benefits of `expl3` that we deal with such things.

I guess you are thinking that `\prop_set:Nnn` would internally check the size of the `prop`. Would that mean having an internal size tracking `int` for performance? Or given we'd switch at a moderate size, can it just use `\prop_count:N`? I think it would be best to avoid an opt-in approach unless the performance of a 'start small and switch' version is really poor.

We've always been clear that people shouldn't rely on implementation: `prop` is one of the things that has changed in the past. I'd say we can adjust here and if someone is using `\tl_set_eq:NN`  then they are on their own.

**Comment by lvjr** [comment]
It may be a drawback that `\prop_set_eq:NN` would be much slower than before.

**Comment by FrankMittelbach** [comment]
I wouldn't track the size I guess. Instead I would leave it up to the programmer if he or she wants large props or small props and have a simple marker at the start that allows to distinguish the two internally. After all it depends probably on the code, e.g. if there are a lot of `\prop_set_eq:N, whether or not you want to use one or the other implementation.

**Comment by josephwright** [comment]
@FrankMittelbach  I can see the idea - if we want to explore that, would you call the new function something like `\prop_new_large:N` then not have any other functions that 'need to worry'?

**Comment by FrankMittelbach** [comment]
yes, just two declarations for new props everything else being transparent. I even wonder if it is worth making double linked lists and doubling the size. My guess is that most of the time you don't do full deletions on props but usually only replacements or additions.

Or one could have internally 3 types: 0 = linear storage as it is now, 1 = single linked list, 2 = double linked list.
A list of type 1 could be transfered to type 2 on the first deletion, given that you have to map over it anyway  in that case.

Alternatively, we could have `\prop_new:nN` with `\prop_new:nN{slink} \g_my_prop`  and `{dlink}` `{inline}` ... and allow to select different implementations.

**Comment by car222222** [comment]
A Rhyme Re:

 "18855 multiletter control sequences out of 15000+600000

  To keep most operations O(1), you can use a doubly linked list:"

One hash to store them all,
(needs) one list to find them.
One link (or two) to bring them all
and for deletion bind them.

(The 'JRR-Strategy' for large PLs?)

**Comment by blefloch** [comment]
I'm starting to implement this, but only two implementation: standard implementation, and doubly-linked list.  Can I use \expanded?  It will make my life somewhat easier, but it is probably not strictly necessary.

**Comment by FrankMittelbach** [comment]
I don't think we should make \expanded "required" earlier than planned  --- so depends a bit on when that is supposed to happen.

**Comment by car222222** [comment]
@blefloch wrote: 
Another drawback is that it slows down mapping by a few additional expansions. 

Comment:
Since any mapping typically involves quite a lot of expansions per item, it seems likely that a few extra will not be a problem -- the process remains O(N): i.e. any Mapping over any large list, whatever the implementation, is in itself already incredibly expensive! Hence such mappings will never a good idea using TeX, I guess.

**Comment by car222222** [comment]
Is there any great need for an “slist” implementation?﻿

Does it have any other advantage beyond saving one token per item?  

Noting that the value part of the item will typically contain several tokens, and it may well contain very many! 

Or are there likely use-cases where saving that one token matters?

**Comment by FrankMittelbach** [comment]
@car222222 you mean single linked lists? For now Bruno said he is only implementing the double-linked version. But to answer your question: dlink lists double the number of burned csnames. It is not the limit in token mem but the limit of available csnames that can still be a problem and what goes into the values in terms of tokens doesn't matter much there because those tokens are not "new".

**Comment by car222222** [comment]
@FrankMittelbach wrote:
dlink lists double the number of burned csnames

Are you sure?  It looks to me as if there is just one csname per item in both implementations.  Where do you see the need for two per item in a dlink list?

**Comment by FrankMittelbach** [comment]
Am 06.02.22 um 14:42 schrieb Chris Rowley:
> Are you sure?

no I'm not sure ... was just talking rubbish, so it is only doubling the 
number of tokens as you say and that is not important

**Comment by car222222** [comment]
Phew!  So I have not totally lost it . . . yet! :-)

**Comment by blefloch** [comment]
I'm actively working on this, it requires a lot of small changes to the code, to make every function check correctly for which type of prop to work with.

I think there will be three new functions, with names to be decided: `\prop_new_large:N`, `\prop_clear_new_large:N`, `\prop_const_large_from_keyval:Nn`.  Perhaps `large` (or any similar description) should come just after `prop`, so `\prop_large_new:N`, as it is almost part of the type's name?

**Comment by car222222** [comment]
I think that prop_large is best.  But if it is a new data-type and not interchangeable then maybe proplarge_ is strictly more correct?

**Comment by FrankMittelbach** [comment]
I also think `prop_large`. But it isn'T really a new datatype it is only that there are two different allocation possibilities, the rest is then transparent. So I think the underscore in the "new" commands is the right thing to do, after that you don't need to know if it is "large" or not.

**Comment by car222222** [comment]
OK, if the two can be treated as a single type for most all operations.  Or rather, as a single type 'after creation':  hoping that this works OK!

**Comment by josephwright** [comment]
@blefloch Any idea on a potential milestone we could set here? Some time in 2023?

**Comment by blefloch** [comment]
For the record I just had the following idea to make `V`-type expansion work for these props as well (for use as items in other props/seq, cf #1067).  Make these long props be `\toks`, and change the "non-expandable" branch of `V`-type expansion to test whether the first token after expanding `\the` is a digit (which happens for the primitive types `int`, `dim`, `skip`, `muskip`) and otherwise call slower code to properly expand the whole doubly-linked list that is stored in a `\toks`.  This requires a bit more internal code to assign toks and make sure that other parts of `expl3` don't overlap with that usage.

**Comment by blefloch** [comment]
@josephwright It seems the main missing step was the code to check validity of a prop (in `\prop_show:N`), which is pretty tricky to do correctly.  Also some further checking that I didn't miss anything.  I'll try to do it this summer before work etc restarts.

**Comment by josephwright** [comment]
We now have the 'large' `prop` structure which parallels the 'classical' approach: I suspect that covers things here. Is that sufficient?
